src/autopas/containers/NeighborListsBuffer.h

```diff
@@ -0,0 +1,148 @@
+      if (index > _lastValidListIndex) {
+        autopas::utils::ExceptionHandler::exception(
+            "NeighborListsBuffer::getNeighborListRef() index out of bounds: {} > {}", index, _lastValidListIndex);
+      }
+    }
+    return _neighborLists[index];
+  }
+
+  /**
+   * Getter for a reference to a neighbor list by key.
+   * @tparam throwIfKeyIsUnknown
+   * @param key
+   * @return
+   */
+  template <bool throwIfKeyIsUnknown = false>
+  std::vector<Value> &getNeighborListRef(const Key &key) {
+    if constexpr (throwIfKeyIsUnknown) {
+      if (_keyMap.find(key) == _keyMap.end()) {
+        autopas::utils::ExceptionHandler::exception("NeighborListsBuffer::getNeighborListRef() unknown key: {}", key);
+      }
+    }
+    return _neighborLists[_keyMap[key]];
+  }
+
+  /**
+   * Reserves a neighbor list for use.
+   * Grows the internal buffer if there are no spare lists left.
+   *
+   * @note This function is NOT thread safe.
+   *
+   * @return
+   */
+  size_t addNeighborList() {
+    // if the buffer is saturated...
+    if (_lastValidListIndex >= _neighborLists.size()) {
+      // ...grow it and initialize all new lists. Make sure to grow to at least 10 lists.
+      _neighborLists.resize(std::max(10ul, static_cast<size_t>(_neighborLists.size() * _growthFactor)),
+                            std::vector<Value>(_defaultListLength));
+    }
+    return ++_lastValidListIndex;
+  }
+
+  /**
+   * Assigns a neighbor list to the given key.
+   *
+   * @note This function is NOT thread safe.
+   *
+   * @param key
+   * @return Index of the neighbor list in the internal data structure.
+   */
+  size_t addNeighborList(const Key &key) {
+    //  size_t addNeighborList(const std::conditional<std::is_pointer_v<Key>, Key, Key &> key) {
+    const auto newIndex = addNeighborList();
+    _keyMap.emplace(key, newIndex);
+    return newIndex;
+  }
+
+  /**
+   * Clears the internal key map and moves _lastValidListIndex to indicate an empty buffer.
+   * This function does not touch the buffer itself.
+   */
+  void clear() {
+    _keyMap.clear();
+    _lastValidListIndex = std::numeric_limits<size_t>::max();
+  }
+
+  /**
+   * Resize the internal buffer so that there are new spare lists.
+   * If n is not larger than the current capacity nothing happens.
+   *
+   * @param n
+   */
+  void reserveNeighborLists(size_t n) { _neighborLists.resize(n, std::vector<Value>(_defaultListLength)); }
+
+  /**
+   * Set the initial length of new neighbor lists.
+   * @param defaultListLength
+   */
+  void setDefaultListLength(size_t defaultListLength) { NeighborListsBuffer::_defaultListLength = defaultListLength; }
+
+  /**
+   * Set the growth factor for the internal buffer.
+   * @param growthFactor
+   */
+  void setGrowthFactor(double growthFactor) { NeighborListsBuffer::_growthFactor = growthFactor; }
+
+ private:
+  /**
+   * Map to associate keys with indexes of _neighborLists.
+   */
+  std::unordered_map<Key, size_t, Hash, KeyEqual> _keyMap{};
```

Comment:
Here? See [the doc of `std::unordered_map`](https://en.cppreference.com/w/cpp/container/unordered_map). Practically speaking, I included this, so that if NeighborListsBuffer is used with a key type that is not trivially hashable, or comparable, you can provide your own stuff the same way as for an `unordered_map`.

Reply:
Ah right, always one step ahead ;)